HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,390 @@
+    bool FailedL2 = false;
+    if (std::abs(L2cand->eta()) > max_Eta_L2_)
+      FailedL2 = true;
+
+    reco::TrackRef L2tk = L2cand->track();
+    LogDebug("HLTMuonL3orL2PreFilter") << " Muon in loop, q*pt= " << L2tk->charge() * L2tk->pt() << " ("
+                                       << L2cand->charge() * L2cand->pt() << ") "
+                                       << ", eta= " << L2tk->eta() << " (" << L2cand->eta() << ") "
+                                       << ", hits= " << L2tk->numberOfValidHits() << ", d0= " << L2tk->d0()
+                                       << ", dz= " << L2tk->dz();
+
+    //normalizedChi2 cut
+    if (L2tk->normalizedChi2() > max_NormalizedChi2_L2_)
+      FailedL2 = true;
+
+    if (not recoBeamSpotHandle.isValid()) {
+      throw cms::Exception("MissingProduct") << "can not find " << beamspotTag_;
+    }  //Throw an exception if the beamSpot is not valid
+
+    if (recoBeamSpotHandle.isValid()) {
+      const reco::BeamSpot& beamSpot = *recoBeamSpotHandle;
+
+      //dz cut (impact parameter of the muon in the z direction computed with respect to the beamspot position)
+      float vz_L2 = L2cand->vz() - beamSpot.z0();
+      float vx_L2 = L2cand->vx() - beamSpot.x0();
+      float vy_L2 = L2cand->vy() - beamSpot.y0();
+      if (std::abs(vz_L2 - (vx_L2 * L2cand->px() + vy_L2 * L2cand->py()) / (L2cand->pt() * L2cand->pt()) *
+                               L2cand->pz()) > max_Dz_)
+        FailedL2 = true;
+
+      // dxy significance cut (safeguard against bizarre values)
+      float const absDxy = std::abs(L2tk->dxy(beamSpot.position()));
+      if (min_DxySig_ > 0 && (L2tk->dxyError() <= 0 || (absDxy < min_DxySig_ * L2tk->dxyError())))
+        FailedL2 = true;
+
+      //dxy beamspot cut
+      if (absDxy > max_DXYBeamSpot_ || absDxy < min_DXYBeamSpot_)
+        FailedL2 = true;
+    }
+
+    //min muon hits cut
+    const reco::HitPattern& L2trackHits = L2tk->hitPattern();
+    if (L2trackHits.numberOfValidMuonHits() < min_NmuonHits_L2_)
+      FailedL2 = true;
+
+    // number of eta bins for cut on number of stations
+    const std::vector<double>::size_type nAbsetaBins = absetaBins_L2_.size();
+
+    // cut on number of stations
+    bool failNstations(false), failNhits(false), failNchambers(false);
+    for (unsigned int i = 0; i < nAbsetaBins; ++i) {
+      if (std::abs(L2cand->eta()) < absetaBins_L2_[i]) {
+        if (L2cand->track()->hitPattern().muonStationsWithAnyHits() < minNstations_L2_[i]) {
+          failNstations = true;
+        }
+        if (L2cand->track()->numberOfValidHits() < min_Nhits_L2_[i]) {
+          failNhits = true;
+        }
+        if ((L2cand->track()->hitPattern().dtStationsWithAnyHits() +
+                 L2cand->track()->hitPattern().cscStationsWithAnyHits() <
+             minNchambers_L2_[i])) {
+          failNchambers = true;
+        }
+        break;
+      }
+    }
+    if (failNstations || failNhits || failNchambers)
+      FailedL2 = true;
```

Comment:
General question: when `FailedL2` is `true`, could you `continue` right away (and go to the next candidate), or do you really need to go through all the following steps in the loop scope?

Reply:
I think it is better to go through. For example, imagine an event with only one muon which is reconstructed in both ways, global and standalone. Because we may have different cuts in L2 and L3 muons, if this muon is rejected in the first loop of L2 muons, then, it will never enter in the second loop.